boot/bootutil/src/image_x509.c

```diff
@@ -108,67 +111,93 @@ bootutil_img_validate(struct enc_key_data *enc_state, int image_index,
              * present.
              */
             if (len != sizeof(hash)) {
-                return -1;
+                ret = -1;
+                goto cleanup;
             }
             rc = flash_area_read(fap, off, buf, sizeof hash);
             if (rc) {
-                return rc;
+                ret = rc;
+                goto cleanup;
             }
             if (memcmp(hash, buf, sizeof(hash))) {
                 BOOT_LOG_ERR("Corrupt hash");
-                return -1;
+                ret = -1;
+                goto cleanup;
             }
 
             sha256_valid = true;
         } else if (type == IMAGE_TLV_X509) {
             if (len > sizeof buf) {
-                return -1;
+                ret = -1;
+                goto cleanup;
             }
             rc = flash_area_read(fap, off, buf, len);
             if (rc) {
-                return rc;
+                ret = rc;
+                goto cleanup;
             }
             // BOOT_LOG_ERR("Cert: %d bytes", len);
-            rc = mbedtls_x509_crt_parse_der(&chain, buf, len);
+            rc =  mbedtls_x509_crt_parse(&chain, buf, len);
             if (rc) {
                 BOOT_LOG_ERR("Parse error %d", rc);
-                /* TODO After init, should cleanup */
-                return -1;
+                ret = -1;
+                goto cleanup;
             }
+
+
         } else if (type == IMAGE_TLV_ECDSA256) {
             /* Finish with the root certificate. */
             mbedtls_x509_crt_init(&trust_ca);
-            rc = mbedtls_x509_crt_parse_der(&trust_ca, bootutil_root_cert,
+            rc =  mbedtls_x509_crt_parse_der(&trust_ca, bootutil_root_cert,
                                             bootutil_root_cert_len);
             if (rc) {
                 BOOT_LOG_ERR("Parser error: %d", rc);
-                /* TODO: After init, should cleanup */
-                return -1;
+                ret = -1;
+                goto cleanup;
             }
 
-            // BOOT_LOG_ERR("Verify chain");
             flags = 0;
             rc = mbedtls_x509_crt_verify(&chain, &trust_ca, NULL,
                                          NULL,
-                                         // "MCUboot sample signing key",
                                          &flags,
                                          verify_callback, NULL);
-            // BOOT_LOG_ERR("Verify chain: %d", rc);
             if (rc == 0) {
-                    cert_valid = true;
-            }
+            	 // validate the image signature
+            	  if (len > sizeof buf) {
+            		  	  	  	ret = -1;
+            		  	  	  	goto cleanup;
+            	            }
+            	            rc = flash_area_read(fap, off, buf, len);
+            	            if (rc) {
```

Comment:
whether you should validate for negative result instead?
`if (rc) {`

Reply:
returns int value 0 for success and 1 for error